service/controller/resource/monitoring/remotewriteagentsecret/resource.go

```diff
@@ -0,0 +1,102 @@
+package remotewriteagentsecret
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/k8sclient/v7/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/prometheus-meta-operator/v2/pkg/password"
+	"github.com/giantswarm/prometheus-meta-operator/v2/service/controller/resource/generic"
+	"github.com/giantswarm/prometheus-meta-operator/v2/service/key"
+)
+
+const (
+	Name = "remotewriteagentsecret"
+)
+
+type Config struct {
+	K8sClient       k8sclient.Interface
+	Logger          micrologger.Logger
+	PasswordManager password.Manager
+}
+
+func New(config Config) (*generic.Resource, error) {
+	clientFunc := func(namespace string) generic.Interface {
+		c := config.K8sClient.K8sClient().CoreV1().Secrets(namespace)
+		return wrappedClient{client: c}
+	}
+
+	c := generic.Config{
+		ClientFunc:    clientFunc,
+		Logger:        config.Logger,
+		Name:          Name,
+		GetObjectMeta: getObjectMeta,
+		GetDesiredObject: func(ctx context.Context, v interface{}) (metav1.Object, error) {
+			return toSecret(ctx, v, config)
+		},
+		HasChangedFunc: hasChanged,
+	}
+	r, err := generic.New(c)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return r, nil
+}
+
+func getObjectMeta(ctx context.Context, v interface{}) (metav1.ObjectMeta, error) {
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return metav1.ObjectMeta{}, microerror.Mask(err)
+	}
+
+	return metav1.ObjectMeta{
+		Name:      key.RemoteWriteAgentSecretName,
+		Namespace: key.Namespace(cluster),
+		Labels:    key.PrometheusLabels(cluster),
+	}, nil
+}
+
+func toSecret(ctx context.Context, v interface{}, config Config) (*corev1.Secret, error) {
+	objectMeta, err := getObjectMeta(ctx, v)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	username := key.ClusterID(cluster)
+
+	password, err := config.PasswordManager.GeneratePassword(32)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	hashedPassword, err := config.PasswordManager.Hash(password)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: objectMeta,
+		Data: map[string][]byte{
+			// To be used by the ingress basic auth
+			"auth":     []byte(fmt.Sprintf("%s:%s", username, hashedPassword)),
```

Comment:
Can we leave this magic to the client/prometheus-agent ? I would rather that we provider username and password and the client use them as he wants to.

Also why are we sending a hashed password here?

Reply:
No we cannot because this auth parameter is used by the ingress on the management cluster and the password needs to be hashed for nginx to work